std/socket.zig

```diff
@@ -0,0 +1,678 @@
+    else => i32,
+};
+
+pub const InvalidSocketFd: SocketFd = switch (builtin.os) {
+    builtin.Os.windows => system.INVALID_SOCKET,
+    else => -1,
+};
+
+pub const Domain = enum {
+    Unspecified,
+    Unix,
+    Inet,
+    Inet6,
+
+    fn toInt(self: Domain, comptime T: type) T {
+        return @intCast(T, switch (self) {
+            Domain.Unspecified => system.AF_UNSPEC,
+            Domain.Unix => system.AF_UNIX,
+            Domain.Inet => system.AF_INET,
+            Domain.Inet6 => system.AF_INET6,
+        });
+    }
+};
+
+pub const SocketType = enum {
+    Stream,
+    DataGram,
+    Raw,
+    SeqPacket,
+
+    fn toInt(self: SocketType, comptime T: type) T {
+        return @intCast(T, switch (self) {
+            SocketType.Stream => system.SOCK_STREAM,
+            SocketType.DataGram => system.SOCK_DGRAM,
+            SocketType.Raw => system.SOCK_RAW,
+            SocketType.SeqPacket => system.SOCK_SEQPACKET,
+        });
+    }
+};
+
+pub const Protocol = enum {
+    Unix,
+    TCP,
+    UDP,
+    IP,
+    IPV6,
+    RAW,
+    ICMP,
+
+    fn toInt(self: Protocol, comptime T: type) !T {
+        return @intCast(T, switch (self) {
+            Protocol.Unix => if (is_posix) 0 else return SocketError.ProtocolNotSupported,
+            Protocol.TCP => if (is_posix) system.PROTO_tcp else system.IPPROTO_TCP,
+            Protocol.UDP => if (is_posix) system.PROTO_udp else system.IPPROTO_UDP,
+            Protocol.IP => if (is_posix) system.PROTO_ip else system.IPPROTO_IP,
+            Protocol.IPV6 => if (is_posix) system.PROTO_ipv6 else system.IPPROTO_IPV6,
+            Protocol.RAW => if (is_posix) system.PROTO_raw else system.IPPROTO_RAW,
+            Protocol.ICMP => if (is_posix) system.PROTO_icmp else system.IPPROTO_ICMP,
+        });
+    }
+};
+
+pub const Shutdown = enum {
+    Read = if (is_posix) system.SHUT_RD else system.SD_RECEIVE,
+    Write = if (is_posix) system.SHUT_WR else system.SD_SEND,
+    Both = if (is_posix) system.SHUT_RDWR else system.SD_BOTH,
+
+    fn toInt(self: Shutdown, comptime T: type) T {
+        return @intCast(T, switch (self) {
+            Shutdown.Read => if (is_posix) system.SHUT_RD else system.SD_RECEIVE,
+            Shutdown.Write => if (is_posix) system.SHUT_WR else system.SD_SEND,
+            Shutdown.Both => if (is_posix) system.SHUT_RDWR else system.SD_BOTH,
+        });
+    }
+};
+
+// pub const Level = enum(u32) {
+//     IP = if (is_posix) system.SOL_IP ,
+//     IPV6 = if (is_posix) system.SOL_IPV6 ,
+//     RM = if (is_posix) system. ,
+//     TCP = if (is_posix) system. ,
+//     UDP = if (is_posix) system. ,
+//     IPX = if (is_posix) system. ,
+//     AppleTalk = if (is_posix) system. ,
+//     IRLMP = if (is_posix) system. ,
+//     Socket = if (is_posix) system. ,
+// };
+
+fn accept4Windows(fd: SocketFd, addr: ?*net.OsAddress, addrlen: ?*system.socklen_t, flags: u32) SocketFd {
+    const result = system.accept(fd, addr, @ptrCast(*c_int, addrlen));
```

Comment:
this `addrlen` pointer cast is not right.  The `accept` function will be assigning it as ` c_int` pointer when it's actually a `socklen_t` pointer which could be a different size.  You'll need to do this:
```zig
var addrlen_cint : c_int = undefined;
accept(...&addrlen_cint...)
addrlen.* = addrlen_cint;
```

Reply:
`system.socklen_t` is defined as a cint for on windows so this should be fine., right?

Reply:
Should be fine if it's the same size. You can add a comptime check to make sure it is. It may not be on all platforms in which case you'll get a compile error if you add the check.